From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Fri, 18 Feb 2022 10:05:45 +0100
Subject: Add IFLA_BR_MCAST_FLOOD

This allows to configure the multicast flooding on a bridge over switchdev
in the kernel. This is used for PON.

Set it like this:
  $ ip link set dev sw257 type bridge mcast_flood 1
---
 include/uapi/linux/if_link.h |  1 +
 ip/iplink_bridge.c           | 15 +++++++++++++++
 2 files changed, 16 insertions(+)

diff --git a/include/uapi/linux/if_link.h b/include/uapi/linux/if_link.h
index 744dd877..ad89b068 100644
--- a/include/uapi/linux/if_link.h
+++ b/include/uapi/linux/if_link.h
@@ -479,6 +479,7 @@ enum {
 	IFLA_BR_VLAN_STATS_PER_PORT,
 	IFLA_BR_MULTI_BOOLOPT,
 	IFLA_BR_MCAST_QUERIER_STATE,
+	IFLA_BR_MCAST_FLOOD,
 	__IFLA_BR_MAX,
 };
 
diff --git a/ip/iplink_bridge.c b/ip/iplink_bridge.c
index d12fd055..dbb8cb8a 100644
--- a/ip/iplink_bridge.c
+++ b/ip/iplink_bridge.c
@@ -59,6 +59,7 @@ static void print_explain(FILE *f)
 		"		  [ mcast_stats_enabled MCAST_STATS_ENABLED ]\n"
 		"		  [ mcast_igmp_version IGMP_VERSION ]\n"
 		"		  [ mcast_mld_version MLD_VERSION ]\n"
+		"         [ mcast_flood (0 == all, 1 == discard, 2 == CPU) ]\n"
 		"		  [ nf_call_iptables NF_CALL_IPTABLES ]\n"
 		"		  [ nf_call_ip6tables NF_CALL_IP6TABLES ]\n"
 		"		  [ nf_call_arptables NF_CALL_ARPTABLES ]\n"
@@ -341,6 +342,14 @@ static int bridge_parse_opt(struct link_util *lu, int argc, char **argv,
 				invarg("invalid mcast_mld_version", *argv);
 			addattr8(n, 1024, IFLA_BR_MCAST_MLD_VERSION,
 				  mld_version);
+		} else if (matches(*argv, "mcast_flood") == 0) {
+			__u8 mcast_flood;
+
+			NEXT_ARG();
+			if (get_u8(&mcast_flood, *argv, 0))
+				invarg("invalid mcast_flood", *argv);
+			addattr8(n, 1024, IFLA_BR_MCAST_FLOOD,
+				  mcast_flood);
 		} else if (matches(*argv, "nf_call_iptables") == 0) {
 			__u8 nf_call_ipt;
 
@@ -655,6 +664,12 @@ static void bridge_print_opt(struct link_util *lu, FILE *f, struct rtattr *tb[])
 			   "mcast_mld_version %u ",
 			   rta_getattr_u8(tb[IFLA_BR_MCAST_MLD_VERSION]));
 
+	if (tb[IFLA_BR_MCAST_FLOOD])
+		print_uint(PRINT_ANY,
+			   "mcast_flood",
+			   "mcast_flood %u ",
+			   rta_getattr_u8(tb[IFLA_BR_MCAST_FLOOD]));
+
 	if (tb[IFLA_BR_NF_CALL_IPTABLES])
 		print_uint(PRINT_ANY,
 			   "nf_call_iptables",
